# Introduction to Argovis's API

Argovis provides an API that indexes and distributes numerous oceanographic datasets with detailed query parameters, enabling you to search and download only and exactly data of interest. In this notebook, we'll tour some of the standard usage patterns enabled by Argovis.

## Setup: Register an API key

In order to allocate Argovis's limited computing resources fairly, users are encouraged to register and request a free API key. This works like a password that identifies your requests to Argovis. To do so:

 - Visit [https://argovis-keygen.colorado.edu/](https://argovis-keygen.colorado.edu/)
 - Fill out the form under _New Account Registration_
 - An API key will be emailed to you shortly.
 
Treat this API key like a password - don't share it or leave it anywhere public. If you ever forget it or accidentally reveal it to a third party, see the same website above to change or deactivate your token.

Put your API key in the quotes in the variable below before moving on:

In [1]:
API_ROOT='https://argovis-api.colorado.edu/'
API_KEY=''

# Argovis data structures

Argovis standard data structures divide measurements into _data_ and _metadata_ documents. Typically, a data document corresponds to measurements or gridded data associated with a discreet temporospatial column - a time, latitude and longitude. A single such document may contain measurements at multiple depths or altitudes, provided they share the same latitude, longitude, and time.

Each of these data documents will refer to at least one corresponding metadata document that captures additional information about the measurement. Argovis divides information between data and metadata documents in order to minimize redundancy in the data you download: many data documents will point to the same metadata document, allowing you to only download that metadata once. Typically, these metadata groupings will refer to some meaningful characteristic of the data; Argo metadata documents correspond to physical floats, while CCHDO metadata documents correspond to cruises, for example.

For more detail and specifications on the data and metadata documents for each collection, see [https://argovis.colorado.edu/docs/documentation/_build/html/database/schema.html](https://argovis.colorado.edu/docs/documentation/_build/html/database/schema.html).

# The standard data routes

## What datasets does Argovis index?

Argovis supports several different data sets with the API and data structures described here. They and their corresponding routes are:

 - Argo profiling float data, `/argo`
 - CCHDO ship-based profile data, `/cchdo`
 - tropical cyclone data from HURDAT and JTWC, `/tc`
 - Global Drifter Program data, `/drifters`
 - several gridded products:
   - Roemmich-Gilson total temperature and salinity, `/grids/rg09`
   - ocean heat content, `/grids/kg21`
   - GLODAP, `/grids/glodap`
 - Argone Argo float position forecast model data, `/argone`
 - Argo trajectory data, `/argotrajectories`
 - several satellite-based timeseries:
   - NOAA sea surface temperature, `/timeseries/noaasst`
   - Copernicus sea surface height, `/timeseries/copernicussla`
   - CCMP wind vector product, `/timeseries/ccmpwind`
   
The examples that follow apply equally to all these routes; they all support similar query options and follow similar behavior patterns.

## Using Swagger and the `argovisHelpers` package to download data

In order to successfully explore Argovis data, there are two important tools to introduce in this section: Swagger, our API documentation engine, and `argovisHelpers`, our Python package of fuctions to help you access and interpret Argovis data.

### Using Swagger docs

Argovis' API documentation is found at [https://argovis-api.colorado.edu/docs/](https://argovis-api.colorado.edu/docs/). These docs are split into several categories; what follows applies to all categories _not_ marked experimental; the experimental categories are under development and may change or be removed at any time.

Categories have three typical routes:
 - The main _data route_, like `/argo`, or `/cchdo`. These routes provide the data documents for the dataset named in the route.
 - The _metadata route_, like `/argo/meta`. These routes provide the metadata documents referred to by data documents.
 - The _vocabulary route_, like `/argo/vocabulary`. These routes provide lists of possible options for search parameters used in the corresponding data and metadata routes.
 
Click on any of the routes, like `/argo` - a list of possible query string parameters are presented, with a short explanation of what they mean.

If you're familiar with REST APIs, this is enough information for you to construct a query string and issue a request in any programming environment that can facilitate an HTTP GET request. If you're working in Python, we provide a helper library, `argovisHelpers`, to manage these requests for you. Let's try it out by making our first request for Argo data, for profiles found within 100 km of a point in the South Atlantic in May 2011 (users of Python's `requests` module will notice a familiar pattern, providing the query string parameters listed in the Swagger docs and associated values as a dictionary):

In [2]:
from argovisHelpers import helpers as avh

In [3]:
argoSearch = {
    'startDate': '2011-05-01T00:00:00Z',
    'endDate': '2011-06-01T00:00:00Z',
    'center': '-22.5,0',
    'radius': 100
}

argoProfiles = avh.query('argo', options=argoSearch, apikey=API_KEY, apiroot=API_ROOT)

Let's have a look at what we get from the first profile returned:

In [4]:
argoProfiles[0]

{'_id': '4901283_003',
 'geolocation': {'type': 'Point', 'coordinates': [-23.139, -0.154]},
 'basin': 1,
 'timestamp': '2011-05-02T08:26:28.000Z',
 'date_updated_argovis': '2023-01-28T01:13:18.370Z',
 'source': [{'source': ['argo_bgc'],
   'url': 'ftp://ftp.ifremer.fr/ifremer/argo/dac/aoml/4901283/profiles/SD4901283_003.nc',
   'date_updated': '2022-06-29T21:21:10.000Z'},
  {'source': ['argo_core'],
   'url': 'ftp://ftp.ifremer.fr/ifremer/argo/dac/aoml/4901283/profiles/D4901283_003.nc',
   'date_updated': '2018-10-03T14:45:37.000Z'}],
 'cycle_number': 3,
 'geolocation_argoqc': 1,
 'profile_direction': 'A',
 'timestamp_argoqc': 1,
 'vertical_sampling_scheme': 'Primary sampling: averaged []',
 'data_info': [['doxy',
   'doxy_argoqc',
   'pressure',
   'pressure_argoqc',
   'salinity',
   'salinity_argoqc',
   'salinity_sfile',
   'salinity_sfile_argoqc',
   'temperature',
   'temperature_argoqc',
   'temperature_sfile',
   'temperature_sfile_argoqc'],
  ['units', 'data_keys_mode'],
  [['

This is a data document for Argo, matching the specification at [https://argovis.colorado.edu/docs/documentation/_build/html/database/schema.html](https://argovis.colorado.edu/docs/documentation/_build/html/database/schema.html). It contains the `timestamp` and `geolocation` properties that place this profile geospatially, and other parameters that typically change from point to point.

All data documents bear a `metadata` key, which is a pointer to the appropriate metadata document to find out more about this measurement. Let's fetch that document for this first profile by querying the `argo/meta` route for a doument with an `id` that matches this `metadata` pointer:

In [5]:
metaOptions = {
    'id': argoProfiles[0]['metadata'][0]
}

argoMeta = avh.query('argo/meta', options=metaOptions, apikey=API_KEY, apiroot=API_ROOT)
argoMeta

[{'_id': '4901283_m0',
  'data_type': 'oceanicProfile',
  'data_center': 'AO',
  'instrument': 'profiling_float',
  'pi_name': ['BRECK OWENS'],
  'platform': '4901283',
  'platform_type': 'SOLO_W',
  'fleetmonitoring': 'https://fleetmonitoring.euro-argo.eu/float/4901283',
  'oceanops': 'https://www.ocean-ops.org/board/wa/Platform?ref=4901283',
  'positioning_system': 'GPS',
  'wmo_inst_type': '851'}]

In addition to temporospatial searches, data and metadata routes typically support _category searches_, which are searches for documents that belong to certain categories. Which categories are available to search by changes logically from dataset to dataset; Argo floats can be searched by platform number, for example, while tropical cyclones can be searched by storm name. See the swagger docs for the full set of possibilities for each category; let's now use argo's platform category search to get all profiles collected by the same platform as the first profile above:

In [6]:
platformSearch = {
    'platform': argoMeta[0]['platform']
}

platformProfiles = avh.query('argo', options=platformSearch, apikey=API_KEY, apiroot=API_ROOT)
print(len(platformProfiles))

125


At the time of writing, 125 profiles are found for this platform in this way.

For all category searches, we may wish to know the full list of all possible values a category can take on; for this, there are the _vocabulary_ routes. Let's get a list of all possible Argo platforms we can search by:

In [7]:
platformVocabSearch = {
    'parameter': 'platform'
}

platforms = avh.query('argo/vocabulary', options=platformVocabSearch, apikey=API_KEY, apiroot=API_ROOT)
print(platforms[0:10])

['13857', '13858', '13859', '15819', '15820', '15851', '15852', '15853', '15854', '15855']


Here we just print out the first 10 platform IDs found, but all 17 thousand or so are present.

## Using the `data` query option

The astute reader may have noticed something about the data document shown above: there's no actual measurements included in it! By default, only the non-measurement data is returned, in order to minimize bandwidth consumed; in order to get back actual measurements and their QC flags, we must query and filter including the `data` parameter, the behavior of which we'll see in this section.

### Basic data request

Let's start by asking for one particular profile by ID, and ask for some temperature data to go with:

In [8]:
dataQuery = {
    'id': '4901283_003',
    'data': 'temperature'
}

profile = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)
print(profile[0]['data'])

[[28.669001, 28.667999, 28.722, 28.816, 28.823, 28.826, 28.830999, 28.783001, 28.775999, 28.740999, 28.694, 28.551001, 28.497, 28.489, 28.414, 28.191999, 28.087999, 28.044001, 27.836, 27.715, 27.655001, 27.41, 27.125999, 26.805, 26.500999, 26.311001, 26.075001, 25.448, 25.120001, 24.632999, 23.709, 22.400999, 21.893, 21.523001, 21.122, 20.861, 20.657, 20.104, 19.707001, 19.681, 19.573999, 19.396999, 19.181, 18.747, 18.438999, 18.240999, 18.045, 17.882, 17.783001, 17.664, 17.563, 17.474001, 17.370001, 17.250999, 17.006001, 16.861, 16.638, 16.465, 16.337999, 16.142, 15.902, 15.721, 15.585, 15.415, 15.28, 15.214, 15.158, 15.084, 15.039, 14.989, 14.91, 14.792, 14.726, 14.698, 14.681, 14.654, 14.618, 14.589, 14.546, 14.459, 14.389, 14.329, 14.219, 14.137, 14.074, 14.039, 14, 13.966, 13.928, 13.851, 13.83, 13.826, 13.823, 13.819, 13.812, 13.808, 13.809, 13.81, 13.808, 13.769, 13.753, 13.736, 13.721, 13.694, 13.657, 13.625, 13.608, 13.584, 13.547, 13.524, 13.508, 13.485, 13.457, 13.418, 13.36

The `data` key contains a list of lists of measurements. To interpret them, we use the `data_info` key:

In [9]:
print(profile[0]['data_info'])

[['temperature', 'pressure'], ['units', 'data_keys_mode'], [['degree_Celsius', 'D'], ['decibar', 'D']]]


`data_info` is always a list that contains exactly three items:

 - `data_info[0]` is the list of measurements returned in our `data` object, in the same order as `data`. So in the example above, `data[0]` are pressure measurements, while `data[1]` are temperature measurements. Note we got back pressures even though we only asked for temperatures; pressures are always provided where available as they are needed to meaningfully interpret all other data variables.
 - `data_info[1]` is a list of per-measurement variables. In the example above, pressure and temperature both have a `units` and a `data_keys_mode` associated with them.
 - `data_info[2]` is a rank 2 matrix with rows labeled by `data_info[0]` and columns by `data_info[1]`. So for the example above, this matrix indicates pressure has units 'decibar', and temperature has `data_keys_mode` 'D'.
 
With this information, we now understand how to interpret the `data` key above: the first list is a list of pressures measured in decibar, and the second list are corresponding temperature measurements measured in degrees C. Note that the ith elements in the data lists all correspond to the same level - in other words, `data[0][i],  data[1][i], data[2][1], ....` are all measurements corresponding to the ith level of this object.
 
> **Data and metadata precedence**: sometimes, you might see a given key on *both* a data document and its corresponding metadata document; when this happens, the value on the data document always takes precedence. `data_info` is a common example of this, which we'll see again below.

### Data inflation

If you find this format difficult to consume, another option is to use the `data_inflate` function from the argovis helpers package. This function will turn your data array into a list of dictionaries, one dictionary per level, with keys corresponding to the data values:

In [10]:
inflated_data = avh.data_inflate(profile[0])
inflated_data[0:10]

[{'temperature': 28.669001, 'pressure': 2},
 {'temperature': 28.667999, 'pressure': 4},
 {'temperature': 28.722, 'pressure': 6},
 {'temperature': 28.816, 'pressure': 8},
 {'temperature': 28.823, 'pressure': 10},
 {'temperature': 28.826, 'pressure': 12},
 {'temperature': 28.830999, 'pressure': 14},
 {'temperature': 28.783001, 'pressure': 16},
 {'temperature': 28.775999, 'pressure': 18},
 {'temperature': 28.740999, 'pressure': 20}]

This format is inefficient to download, but easy to read and work with. Long-time users of previous versions of Argovis may recognize this as similar to the legacy format of some of our data.

### Getting absolutely everything

What we've seen above allows us to be very targeted in the data we download; rather than being forced to spend time and bandwidth downloading data we aren't interested in, we can focus on just what we need. On the other hand, somtimes we really do want everything, and for that there's `data=all`:

In [11]:
dataQuery = {
    'id': '4901283_003',
    'data': 'all'
}

profile = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)
avh.data_inflate(profile[0])[0:10]

[{'doxy': None,
  'doxy_argoqc': 4,
  'pressure': 2,
  'pressure_argoqc': 1,
  'salinity': 35.574966,
  'salinity_argoqc': 1,
  'salinity_sfile': 35.574966,
  'salinity_sfile_argoqc': 1,
  'temperature': 28.669001,
  'temperature_argoqc': 1,
  'temperature_sfile': 28.669001,
  'temperature_sfile_argoqc': 1},
 {'doxy': None,
  'doxy_argoqc': 4,
  'pressure': 4,
  'pressure_argoqc': 1,
  'salinity': 35.573761,
  'salinity_argoqc': 1,
  'salinity_sfile': 35.573761,
  'salinity_sfile_argoqc': 1,
  'temperature': 28.667999,
  'temperature_argoqc': 1,
  'temperature_sfile': 28.667999,
  'temperature_sfile_argoqc': 1},
 {'doxy': None,
  'doxy_argoqc': 4,
  'pressure': 6,
  'pressure_argoqc': 1,
  'salinity': 35.626602,
  'salinity_argoqc': 1,
  'salinity_sfile': 35.626602,
  'salinity_sfile_argoqc': 1,
  'temperature': 28.722,
  'temperature_argoqc': 1,
  'temperature_sfile': 28.722,
  'temperature_sfile_argoqc': 1},
 {'doxy': None,
  'doxy_argoqc': 4,
  'pressure': 8,
  'pressure_argoqc': 1,

> **Downloading only what you need:**
Some objects, like Argo BGC probes, measure many values. Your downoads will often be dramatically faster if you specify your variables of interest, rather than using `data=all` unnecessarily. Recall that the `data` parameter can also accept a comma-separated list of variable names, if there are a few that you'd like.

### Filtering behavior of data requests

Note that adding a specific data filter is a _firm requirement_ that all returned profiles have some meaningful data for _all_ variables listed. Try demanding chlorophyl-a in addition to temperature for our current profile of interest:

In [12]:
dataQuery = {
    'id': '4901283_003',
    'data': 'temperature,chla'
}

profile = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)
print(profile)

[]


We get nothing in our array of profiles; even though we asked for profile id '4901283_003' and we know it exists, `data=temperature,chla` filters our query down to _only_ profiles that have both temperature and chla reported; since the profile requested doesn't have any chla measurements, it is dropped from the returns in this case. This is useful if you only want to download profiles that definitely have data of interest; for example, try the same thing on our regional search from above:

In [13]:
argoSearch = {
    'startDate': '2011-05-01T00:00:00Z',
    'endDate': '2011-06-01T00:00:00Z',
    'center': '-22.5,0',
    'radius': 100,
    'data': 'temperature,chla'
}

argoProfiles = avh.query('argo', options=argoSearch, apikey=API_KEY, apiroot=API_ROOT)
print(len(argoProfiles))

0


Evidently Argo made no chlorophyl-a measurements in May 2011 within 100 km of our point of interest - a fact which we found using the data api without having to download or reduce any data at all. One final point on data filtering in this manner: it's not enough for a profile to nominally have a variable defined for it; it must have at least one non-null value reported for that variable somewhere in the search results. For example, when we did `data=all` for our profile of interest above, we saw dissolved oxygen, `doxy`, was defined for it. But:

In [14]:
dataQuery = {
    'id': '4901283_003',
    'data': 'doxy'
}

profile = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)
print(profile)

[]


Again our search is filtered down to nothing, since every level in that profile reported `None` for `doxy`.

### Search negation

Let's find some profiles that do actually have dissolved oxygen in them, this time with a slightly different geography search: let's look for everything in August 2017 within a polygon region, defined as a list of `[longitude, latitude]` points: 

In [15]:
dataQuery = {
    'startDate': '2017-08-01T00:00:00Z',
    'endDate': '2017-09-01T00:00:00Z',
    'polygon': [[-150,-30],[-155,-30],[-155,-35],[-150,-35],[-150,-30]],
    'data': 'doxy'
}

profiles = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)

We find one profile with meaningful dissolved oxygen data in the region of interest.

The `data` key also accepts _tilde negation_, meaning 'filter for profiles that _don't_ contain this data', for example:

In [16]:
dataQuery = {
    'startDate': '2017-08-01T00:00:00Z',
    'endDate': '2017-09-01T00:00:00Z',
    'polygon': [[-150,-30],[-155,-30],[-155,-35],[-150,-35],[-150,-30]],
    'data': 'temperature,~doxy'
}

profiles = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)
print(len(profiles))

0


We get a collection of profiles that appear in the region of interest, and have temperature but _not_ dissolved oxygen. In this way, we can split up our downloads into groups of related and interesting profiles without re-downloading the same profiles over and over.

### QC filtering

In addition to querying and filtering by what data is available, we can also make demands on the quality of that data by performing QC filtering. Let's start by looking at some particulate backscattering data:

In [17]:
bbpQuery = {
    'id': '2902857_001',
    'data': 'bbp700,bbp700_argoqc'
}

bbp = avh.query('argo', options=bbpQuery, apikey=API_KEY, apiroot=API_ROOT)
bbpindex = bbp[0]['data_info'][0].index('bbp700')
bbpQCindex = bbp[0]['data_info'][0].index('bbp700_argoqc')
print(bbp[0]['data'][bbpindex][0:10])
print(bbp[0]['data'][bbpQCindex][0:10])

[0.004452, 0.004286, 0.004157, 0.004542, 0.005183, 0.004568, 0.004465, 0.005632, 0.005568, 0.00426]
[1, 1, 1, 1, 1, 1, 4, 1, 1, 4]


We request both the measurement and its corresponding QC flags, for reference. Recall that for Argo:

 - QC 1 == definitely good data
 - QC 2 == probably good data
 - QC 3 == probably bad data
 - QC 4 == definitely bad data
 
If we didn't look at the QC flags for our particulate backscatter data, we could easily have missed that some of the measurements shown above (and many more in the profile not printed) have been marked as bad data by the upstream data distributor, and therefore might not be appropriate for your purposes. We can suppress measurements based on a list of allowed QC values by modifying what we pass to the `data` query parameter:

In [18]:
bbpQCfilteredQuery = {
    'id': '2902857_001',
    'data': 'bbp700,1,bbp700_argoqc'
}

bbpFiltered = avh.query('argo', options=bbpQCfilteredQuery, apikey=API_KEY, apiroot=API_ROOT)
bbpindex = bbpFiltered[0]['data_info'][0].index('bbp700')
bbpQCindex = bbpFiltered[0]['data_info'][0].index('bbp700_argoqc')
print(bbpFiltered[0]['data'][bbpindex][0:10])
print(bbpFiltered[0]['data'][bbpQCindex][0:10])

[0.004452, 0.004286, 0.004157, 0.004542, 0.005183, 0.004568, None, 0.005632, 0.005568, None]
[1, 1, 1, 1, 1, 1, 4, 1, 1, 4]


In our `data` query parameter, we listed which QC flags we find tolerable for each measurement parameter; in this case `bbp700,1` indicates we only want `bbp700` data if it has a corresponding QC flag of 1. Some things implied by this example that are worth highlighting:

 - QC flags listed after a variable name only apply to that variable name. Try printing the `pressure` record for the profile found above, and you'll see none of its levels were suppressed.
 - The list of QC flags is an explicit-allow list and can contain as many flags as you want. For example, you might change the above data query to `bbp700,1,2` to get both 1- and 2-flagged `bbp700` measurements back.
 - We include the explicit QC flag in this example for illustrative purposes, but it's not required when doing QC filtering in this way. Try the above query while omitting `bbp700_argoqc`, and you'll get the same non-`None` values for `bbp700`.
 - Note however, as with all data requests, if all explicitly requested data variables are `None` for a level, that level is dropped. In the case where you omitted `bbp700_argoqc` and only requested `bbp700`, the levels where the QC filtration set the `bbp700` value to `None` are dropped.
 - Similarly, if *all* levels of a requested variable are set to `None` by QC filtration, the entire profile will be dropped from the returns, on the grounds that it doesn't contain any of the data you requested at a level of quality you marked as acceptable.
 
Note that QC flags are currently only available for the argo and cchdo datasets, and furthermore that these datasets assign different meanings to their QC flags. Be sure to check the docs for each project to make sure you understand how to interpret that project's QC flags.

### Minimal data responses

Sometimes, we might want to use the `data` filter as we've seen to confine our attention to only profiles that have data of interest, but we're only interested in general or metadata about those measurements, and don't want to download the actual measurements; for this, we can add the `except-data-values` token:

In [19]:
dataQuery = {
    'startDate': '2017-08-01T00:00:00Z',
    'endDate': '2017-09-01T00:00:00Z',
    'polygon': [[-150,-30],[-155,-30],[-155,-35],[-150,-35],[-150,-30]],
    'data': 'doxy,except-data-values'
}

profiles = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)
print(profiles)

[{'_id': '5905107_001', 'geolocation': {'type': 'Point', 'coordinates': [-154.974, -32.415]}, 'basin': 2, 'timestamp': '2017-08-11T11:57:19.001Z', 'date_updated_argovis': '2023-01-27T08:37:16.397Z', 'source': [{'source': ['argo_bgc'], 'url': 'ftp://ftp.ifremer.fr/ifremer/argo/dac/aoml/5905107/profiles/SD5905107_001.nc', 'date_updated': '2022-07-09T07:14:33.000Z'}, {'source': ['argo_core'], 'url': 'ftp://ftp.ifremer.fr/ifremer/argo/dac/aoml/5905107/profiles/D5905107_001.nc', 'date_updated': '2019-06-24T15:29:23.000Z'}], 'cycle_number': 1, 'geolocation_argoqc': 1, 'profile_direction': 'A', 'timestamp_argoqc': 1, 'vertical_sampling_scheme': 'Primary sampling: mixed [deeper than nominal 985dbar: discrete; nominal 985dbar to surface: 2dbar-bin averaged]', 'data_info': [['doxy', 'pressure'], ['units', 'data_keys_mode'], [['micromole/kg', 'D'], ['decibar', 'D']]], 'metadata': ['5905107_m0']}]


Note that specifying only `'data': 'except-data-values'` is the same as just leaving the `data` query key off completely; the purpose of this option is to allow you to filter by data, but then only get back the lightweight non-measurement values. 

If we want an even more minimal response, we can use the `compression=minimal` option:

In [20]:
dataQuery = {
    'startDate': '2017-08-01T00:00:00Z',
    'endDate': '2017-09-01T00:00:00Z',
    'polygon': [[-150,-30],[-155,-30],[-155,-35],[-150,-35],[-150,-30]],
    'data': 'doxy',
    'compression': 'minimal'
}

profiles = avh.query('argo', options=dataQuery, apikey=API_KEY, apiroot=API_ROOT)
print(profiles)

[['5905107_001', -154.974, -32.415, '2017-08-11T11:57:19.001Z', ['argo_bgc', 'argo_core'], ['5905107_m0']]]


With `compression: minimal`, for each data document we get only a minimal amount of information describing it; each data product has a slightly different minimal representation tailored to suit.

### Temporospatial request details

You have seen in examples above that requests can be temporally limited by `startDate` and `endDate`, and confined to a geographic region with `polygon`. There are a few more features and facts about temporospatial requests in Argovis that are worth exploring.

#### Very large spatial extents

Argovis uses geojson polygons to define spatial regions of interest, as illustrated above. If we consider only vertexes, ambiguity exists: are we describing the portion of the globe on one side of the polygon line, or the other? By default, MongoDB and Argovis assume you are asking for the smaller of the two regions. If in fact you want the larger, we must use the winding order of points in the polygon to disambiguate. If we want the larger region, we must tell the API to consider winding order, and make sure the polygon vertexes are listed in counter-clockwise order around our region of interest, like so:

In [21]:
# small square region with CW winding == complimentary region with CCW winding
winding = {
    'startDate': '2023-01-01T00:00:00Z',
    'endDate':   '2023-01-10T00:00:00Z',
    'polygon':   [[-40,35], [-40,45], [-30,45], [-30,35], [-40,35]],
    'winding':   'true',
    'compression': 'minimal'
}

profiles = avh.query('argo', options=winding, apikey=API_KEY, apiroot=API_ROOT)
print(len(profiles))

4046


Note that if we leave off the `winding: 'true'` part, we get the smaller region regardless of winding:

In [22]:
winding = {
    'startDate': '2023-01-01T00:00:00Z',
    'endDate':   '2023-01-10T00:00:00Z',
    'polygon':   [[-40,35], [-40,45], [-30,45], [-30,35], [-40,35]],
    'compression': 'minimal'
}

profiles = avh.query('argo', options=winding, apikey=API_KEY, apiroot=API_ROOT)
print(len(profiles))

9


Or, if we keep the winding requirement but reverse the winding of `polygon`, again we get the smaller region:

In [23]:
winding = {
    'startDate': '2023-01-01T00:00:00Z',
    'endDate':   '2023-01-10T00:00:00Z',
    'polygon':   [[-40,35], [-30,35], [-30,45], [-40,45],  [-40,35]],
    'winding':   'true',
    'compression': 'minimal'
}

profiles = avh.query('argo', options=winding, apikey=API_KEY, apiroot=API_ROOT)
print(len(profiles))

9


#### Region intersections

In the event we are curious to see results that are interior to the intersection of two or more polygons, Argovis also presents a `multipolygon` option:

In [24]:
poly_intersection = {
    'startDate': '2023-01-01T00:00:00Z',
    'endDate':   '2023-01-10T00:00:00Z',
    'multipolygon':   [
        [[-40,35], [-40,45], [-30,45], [-30,35], [-40,35]],
        [[-35,35], [-35,45], [-25,45], [-25,35], [-35,35]]
    ],
    'compression': 'minimal'
}

profiles = avh.query('argo', options=poly_intersection, apikey=API_KEY, apiroot=API_ROOT)
print(len(profiles))

6


Note the profiles returned in this case are interior to _both_ polygons listed. There's no limit to the number of polygons you can list in `multipolygon`; each will further filter down the number of profiles returned to be interior to all of them.